spec/protocol/delegation.adoc

```diff
@@ -0,0 +1,90 @@
+
+[plantuml, format=png]
+....
+@startuml
+title Delegation Request Sequence
+
+actor alice as "Alice\nmabc..defg"
+actor bob as "Bob\nmhij..klmn"
+participant server as "Server"
+
+== Certificate Creation ==
+bob -> alice : Public Key
+alice -> alice : `many delegate create`
+alice -> bob : certificate
+
+== Request ==
+
+bob -> server : message\n\
+    signature: mhij..klmn\n\
+    from: mabc..defg\n\
+    attributes: [2, certificate]
+activate server
+server -> bob : Response
+deactivate server
+
+@enduml
+....
+
+=== Attributes
+
+There are 3 attributes related to this specification:
+
+- xref:../../attributes/network/10_delegation.adoc[10 – Delegation] which allows a server to sign delegations (if supported) and advertise they support delegation.
+- xref:../../attributes/request/2_delegation.adoc[2 - Delegation (Request)] which allows requests to have a delegation.
+- xref:../../attributes/response/2_delegation.adoc[2 - Delegation (Response)] which allows responses to have a delegation.
+
+The certificate is described in the spec CDDL and referred by the attributes above.
+Certificates are passed along with a request or response as an attribute, and are not part of the envelope.
+
+=== Acceptance
+
+A server who does not advertise xref:../../attributes/network/10_delegation.adoc[attribute 10] MUST error out if the attribute 2 is found in a request (see <<./attributes.adoc#_unknown_attributes,Unknown Attributes>>).
+
+=== Anonymous
+
+Please note that even though the anonymous address does not have a public key, it is a valid identity and can be delegated _to_, as well as from, although it has no utility.
+
+== Resolution
+
+A server supporting attribute 10 MUST add a resolution step 3 (see <<./envelope.adoc#_identity,Identity Resolution>>) that does the following:
+
+1. look for attribute `2` in the `message`;
+   a. if the attribute is not found, the server MUST use the previous resolved identity.
+   b. if the attribute is found, read its argument and;
+      . if the argument is an array of certificates, perform step 2 and 3 for each certificate, in order,
+      . if the argument is a single certificate, perform step 2 and 3 once for it.
+
+2. for each certificate, validate (in order);
+   a. the signature using the same <<./envelope.adoc#_identity,Identity Resolution>> steps for a regular envelope,
+   b. the `to` (1) field is the same as the identity resolved thus far (either the last identity resolved as part of this step, or the identity resolved from the message envelope),
+   c. the `expiration` (2) field is still in the future (must also not be equal).
+   d. if the `final` (4) field is `true`, and there are more certificates to evaluate, the server MUST return an error.
+
+3. if all validation above pass, the latest identity is the `from` (0) field of the certificate.
+   If there are more certificate to resolve, go back to step 2 with the next certificate.
```

Comment:
I removed the last line since at step 2 I already say "for each certificate". I could provide pseudocode but I don't think a diagram would be easy to make in a way that makes it also useful.

Reply:
Pseudocode would be a good compromise.